scripts/nbd-firewall-config.sh

```diff
@@ -0,0 +1,91 @@
+#!/bin/bash
+#
+# Copyright (c) Citrix Systems. All rights reserved.
+
+##############################################################
+# Use this script to open/close the specified interfaces
+# (IP addresses) for incoming TCP connections to the NBD port.
+#
+# Usage:
+#   ./nbd-firewall-config set [address...]
+#
+##############################################################
+
+# Historical note: the history of this file has a version that
+# filters by local IP addresses rather than by network interfaces;
+# that version might one day be a useful reference for something.
+
+# Require the "set" argument because otherwise there would be a risk
+# of someone running the script with no arguments and accidentally
+# setting NBD to be available on no addresses.
+# This way, omitting "set" (or supplying "--help") causes the script
+# to print usage instructions.
+
+set -ex
+OP="$1"
+if [ _"${OP}" != _set ]; then
+    echo "Usage: $(basename ${0}) set [interface...]
+will set the firewall to allow incoming TCP connections
+to the NBD port on only the specified interfaces (if any)." 1>&2
+    exit 1
+fi
+shift 1
+
+set -eux
+# No automatic cleanup on error because it would depend too
+# critically on the stage at which the error occurs.
+
+NBDPORT=10809
+# Note that chain names must be under 29 characters.
+NEWCHAININ=xapi_nbd_input_chain_new
+NBDCHAININ=xapi_nbd_input_chain
+NEWCHAINOUT=xapi_nbd_output_chain_new
+NBDCHAINOUT=xapi_nbd_output_chain
+
+# Rule fragments for sending NBD packets to NBD chains
+# from INPUT and OUTPUT chains.
+NBD_INPUT_JUMP_TO="-p tcp --dport $NBDPORT -j "
+NBD_OUTPUT_JUMP_TO="-p tcp --sport $NBDPORT -j "
+
+# Same principle as double-buffering a display:
+# add the items one by one to temporary chains,
+# then swap them into use atomically, or as near as possible.
+iptables --new-chain $NEWCHAININ 2>/dev/null || iptables --flush $NEWCHAININ
+iptables --new-chain $NEWCHAINOUT 2>/dev/null || iptables --flush $NEWCHAINOUT
+while [ "$#" -ne 0 ]; do
+    intf="${1}"
+    # (Each positional param is forgotten for ever at the end of the loop.)
+    shift 1
+    iptables -A $NEWCHAININ -m conntrack --ctstate NEW,ESTABLISHED --in-interface "${intf}" -j ACCEPT
+    iptables -A $NEWCHAINOUT --out-interface "${intf}" -j RETURN
+done
+iptables --append $NEWCHAININ -j REJECT
+iptables --append $NEWCHAINOUT -j REJECT
+
+function insert_rule_if_absent {
+    chain="${1}"
+    shift 1
+    rule="${*}"
+    if ! iptables --check "${chain}" $rule 2>/dev/null
```

Comment:
Why we are not using `[ ]` to enclose the check?
And wouldn't it be an extra safety to add quotes to the rule array?

Reply:
What do you mean about enclosing the check?
`[ ... ]` is an alternative syntax for the `test` bash builtin command, which we are not using here.
Or if you're referring to the `--check`, that is an iptables command which takes the same parameters as the `--insert` and `--append` commands (and one form of the `--delete` command) and just checks whether a rule matching the specification exists. See the man-page for iptables.

I think if I were going to change this at all, I would change it to use `... ||` instead of `if ! ... then`
```
iptables --check "${chain}" $rule 2>/dev/null \
    || iptables --insert "${chain}" $rule
```
Obviously, the aim is to insert the rule into the chain iff it is not there already.

As for writing `"$rule"` (or `"${rule}"`) instead of `$rule`, no, that wouldn't work because the `iptables` command expects to see the rule as multiple separate arguments, not one big one. Look at this little demonstration:
```
[root@xrtuk-01-01 ~]# rule='a b c'
[root@xrtuk-01-01 ~]# echo $rule
a b c
[root@xrtuk-01-01 ~]# function countargs {
> echo "$#"
> }
[root@xrtuk-01-01 ~]# countargs $rule
3
[root@xrtuk-01-01 ~]# countargs "$rule"
1
```

Reply:
I see, thanks for the explanation. Personally I find your suggested change more readable:
```bash
iptables --check "${chain}" $rule 2>/dev/null \
    || iptables --insert "${chain}" $rule
```

Reply:
I've changed from `if-then` to shortcutting `||` to control the flow here.